### PR TITLE
awssd SRV support

### PR DESCRIPTION
**Pull request by SimonePomata**
Internal PR for collaboration, WIP.

### CODE REVIEW

File: provider/awssd/aws_sd.go [hunk 1]
```diff
@@ -218,24 +265,29 @@ func (p *AWSSDProvider) ApplyChanges(ctx context.Context, changes *plan.Changes)
...
-	// Deletes must be executed first to support update case.
-	// When just list of targets is updated `[1.2.3.4] -> [1.2.3.4, 1.2.3.5]` it is translated to:
-	// ```
-	// deletes = [1.2.3.4]
-	// creates = [1.2.3.4, 1.2.3.5]
-	// ```
-	// then when deletes are executed after creates it will miss the `1.2.3.4` instance.
-	err = p.submitDeletes(namespaces, changes.Delete)
+	// remove redundant targets from Create and Delete
+	dedupCreate, dedupDelete, err = p.handleSRVChanges(namespaces, dedupCreate, dedupDelete)
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: The comment seems to be a copy&paste

File: provider/awssd/aws_sd.go [hunk 2]
```diff
@@ -682,3 +1164,91 @@ func (p *AWSSDProvider) isAWSLoadBalancer(hostname string) bool {
 
 	return matchElb || matchNlb
 }
+
+// determine if a given service is of type A+SRV
+func (p *AWSSDProvider) isAAndSRVService(srv *sd.Service) bool {
+	return len(srv.DnsConfig.DnsRecords) == 2 && aws.StringValue(srv.DnsConfig.DnsRecords[0].Type) == sd.RecordTypeA && aws.StringValue(srv.DnsConfig.DnsRecords[1].Type) == sd.RecordTypeSrv
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: This may not always work, the Cloud Map API doesn't guarantee any particular order of `DnsRecords` in the list.

File: provider/awssd/aws_sd.go [hunk 3]
```diff
@@ -530,20 +888,137 @@ func (p *AWSSDProvider) DeregisterInstance(service *sd.Service, ep *endpoint.End
...
+	}
+	log.Infof("Deleting a service \"%s\" (%s)", *svc.Name, *svc.Id)
+	if !p.dryRun {
+		_, err := p.client.DeleteService(&sd.DeleteServiceInput{
+			Id: svc.Id,
+		})
+		if err != nil {
+			return err
+		}
+	}
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: This part fails in the real test. The issue is that the service cannot be deleted while instances are being removed.
```
ERRO[1843] ResourceInUse: Service contains registered instances; delete the instances before deleting the service
```

File: provider/awssd/aws_sd.go [hunk 4]
```diff
@@ -218,24 +265,29 @@ func (p *AWSSDProvider) ApplyChanges(ctx context.Context, changes *plan.Changes)
 	changes.Delete = append(changes.Delete, deletes...)
 	changes.Create = append(changes.Create, creates...)
 
+	// remove redundant targets from Create and Delete
+	dedupCreate, dedupDelete, err := p.DedupDeletesAndCreates(changes.Create, changes.Delete)
+	if err != nil {
+		return err
+	}
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: Maybe this part is not needed. Duplicated actions are created just a couple of lines above with `updatesToCreates`. Isn't it better to modify that instead of removing duplicates later?

File: provider/awssd/aws_sd.go [hunk 5]
```diff
@@ -264,6 +316,220 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
 	return creates, deletes
 }
 
+func (p *AWSSDProvider) handleSRVChanges(namespaces []*sd.NamespaceSummary, creates []*endpoint.Endpoint, deletes []*endpoint.Endpoint) ([]*endpoint.Endpoint, []*endpoint.Endpoint, error) {
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go:325]: Could this function be broken down into smaller pieces, please? It feels like all the magic happens here in `200` lines and it's a bit difficult to follow. It would be nice to have a description of what this function returns.

File: provider/awssd/aws_sd.go [hunk 6]
```diff
@@ -264,6 +316,220 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
...
+		}
+
+		services, err := p.ListServicesByNamespaceID(aws.String(nsID))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		//create a map of relevant changes by host. will loop this map later to determine any SRV-related changes
+		changesByHost := make(map[string]*relevantChanges)
+		if okNsCre {
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: This condition is redundant. It should be safe to iterate an empty slice.

File: provider/awssd/aws_sd.go [hunk 7]
```diff
@@ -264,6 +316,220 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
...
+					chByHost.SRVDel = append(chByHost.SRVDel, ch)
+				}
+			}
+		}
+
+		// loop the map of relevant changes by host
+		for host, relCh := range changesByHost {
+			_, svcName := p.parseHostname(host)
+
+			// defome some relevant variables for handling the host-level changes
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: Typo in the comment

File: provider/awssd/aws_sd.go [hunk 8]
```diff
@@ -264,6 +316,220 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
...
+					host := ch.DNSName
+					if _, ok := changesByHost[host]; !ok {
+						changesByHost[host] = &relevantChanges{}
+					}
+					chByHost := changesByHost[host]
+					if len(chByHost.ACre) == 0 {
+						chByHost.ACre = append(chByHost.ACre, ch)
+					} else {
+						log.Warnf("Skipping record %s because only 1 create change is expected for A record. This is unexpected.", ch.String())
+					}
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: I believe it requires a high level comment describing the data transformation. The code is quite complex.

File: provider/awssd/aws_sd.go [hunk 9]
```diff
@@ -264,6 +316,220 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
...
+				if err != nil {
+					return nil, nil, err
+				}
+			}
+
+			// set the outputs of the main if-block
+			ASvcRemove := false
+			newPrefSRV := ""
+
+			//run the main if-block
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: Those comments just show that the code need to be broken down to smaller pieces. The `if-else` branching became too complex and it's hard to follow.

**Review by vanekjar — CHANGES_REQUESTED** [review]

File: provider/awssd/aws_sd.go (hunk 4 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: Done. I still keep `DedupDeletesAndCreates` as a separate function from `updatesToCreates` in order to improve readibility. But now the former function is invoked from within the latter, and is a lot simpler.

File: provider/awssd/aws_sd.go (hunk 1 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: fixed

File: provider/awssd/aws_sd.go (hunk 6 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: fixed

File: provider/awssd/aws_sd.go (hunk 8 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: added a higher level description

File: provider/awssd/aws_sd.go (hunk 7 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: fixed

File: provider/awssd/aws_sd.go (hunk 3 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: Yes, I know that it could error. I left it there in order to avoid an additional if (and make the code more readeable). Eventually, the service will be deleted during the next iteration anyway. I added an if to avoid the error.

File: provider/awssd/aws_sd.go (hunk 2 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: handled. also created a new function `mainDnsRecord` to use when previously using `srv.DnsConfig.DnsRecords[0]`

File: provider/awssd/aws_sd.go (hunk 5 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go:325]: broken into 3 logical pieces

File: provider/awssd/aws_sd.go (hunk 9 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: done, simplified and added detailed comments

File: provider/awssd/aws_sd.go [hunk 10]
```diff
@@ -11,13 +11,14 @@ Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
-limitations under the License.
+limitations under the License..
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: ?

File: provider/awssd/aws_sd.go [hunk 11]
```diff
@@ -252,18 +306,273 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
...
+
+		services, err := p.ListServicesByNamespaceID(aws.String(nsID))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		//create a map of relevant changes by host. It contains the following endpoints: A create, SRV create, and SRV delete.
+		// A & SRV endpoints are registered as a unified A+SRV service in Cloud Map. Therefore, this map is needed to evaluate changes to the A+SRV service.
+		changesByHost := make(map[string]*SRVRelevantChanges)
+		err = p.SRVPopulateChangesByHost(changesByHost, nsCreates, nsDeletes)
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: Why does this function return output through an argument? Isn't it better to return `changesByHost`?

File: provider/awssd/aws_sd.go [hunk 12]
```diff
@@ -252,18 +306,273 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
...
+				return err
+			}
+			if _, ok := changesByHost[host]; !ok {
+				changesByHost[host] = &SRVRelevantChanges{}
+			}
+			chByHost := changesByHost[host]
+			chByHost.SRVDel = append(chByHost.SRVDel, ch)
+		}
+	}
+	return nil
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: This is confusing. It looks like the function is doing nothing, but it's not true.

File: provider/awssd/aws_sd.go [hunk 13]
```diff
@@ -252,18 +306,273 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
...
+	return newEndpoint
+}
+
+// this function determines the actions to be taken on an SRV record
+func (p *AWSSDProvider) SRVDetermineActions(host string, relCh *SRVRelevantChanges, services map[string]*sd.Service) (ASvcExists bool, ASvcRemove bool, ACreExists bool, ASvcIsAAndSRV bool,
+	ASvc *sd.Service, ACre *endpoint.Endpoint, ASvcCurrentPrefSRV string, newPrefSRV string, err error) {
+
+	_, svcName := p.parseHostname(host)
+	// define some relevant input variables for handling the host-level changes
+	// based on these variables, we
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: typo

File: provider/awssd/aws_sd.go [hunk 14]
```diff
@@ -83,6 +95,13 @@ type AWSSDProvider struct {
 	namespaceTypeFilter *sd.NamespaceFilter
 }
 
+// a struct of all related relevant changes to handle SRV endpoints.
+type SRVRelevantChanges struct {
+	ACre   []*endpoint.Endpoint // max len is 1. This is a create endpoint with recordType A
+	SRVCre []*endpoint.Endpoint // max len is 1. This is a delete endpoint with recordType SRV
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: Why is this defined as a slice when `len <= 1`?

File: provider/awssd/aws_sd.go [hunk 15]
```diff
@@ -274,9 +583,20 @@ func (p *AWSSDProvider) submitCreates(namespaces []*sd.NamespaceSummary, changes
 		}
 
 		for _, ch := range changeList {
+			// skip SRV endpoints here. SRV endpoints are handled as part of A endpoints, and registered under an A+SRV service.
+			// skip endpoints with empty DNSName. These are empty endpoints, likley coming from SRV handleSRVChanges
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go:586]: This comment may be obsolete

File: provider/awssd/aws_sd.go [hunk 16]
```diff
@@ -682,3 +1177,113 @@ func (p *AWSSDProvider) isAWSLoadBalancer(hostname string) bool {
...
+	}
+	parts := strings.Split(target, " ")
+	port = parts[2]
+	host = parts[3]
+	prio = parts[0]
+	weight = parts[1]
+	return
+}
+
+func (p *AWSSDProvider) preferredSRVCombine(hostname string, ttl endpoint.TTL, target string) (prefSRV string, err error) {
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go]: I struggle to understand the keyword `preferred`. What is the meaning in this context?

File: provider/awssd/aws_sd.go [hunk 17]
```diff
@@ -252,18 +306,273 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
...
 		current := updateNewMap[old.DNSName]
 
+		// always register (or re-register) instance with the current data
+		creates = append(creates, current)
+
 		if !old.Targets.Same(current.Targets) {
 			// when targets differ the old instances need to be de-registered first
 			deletes = append(deletes, old)
+			// remove from both current.Targets and create.Targets the targets that appear in both of them
+			p.DedupDeletesAndCreates(current, old)
```
Comment (vanekjar) [inline provider/awssd/aws_sd.go:316]: How does this happen? Shouldn't `current` and `old` be always different when it's update?

**Review by vanekjar — CHANGES_REQUESTED** [review]

File: provider/awssd/aws_sd.go (hunk 10 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: typo, fixed.

File: provider/awssd/aws_sd.go (hunk 11 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: done

File: provider/awssd/aws_sd.go (hunk 12 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: fixed

File: provider/awssd/aws_sd.go (hunk 13 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: fixed

File: provider/awssd/aws_sd.go (hunk 14 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: fixed

File: provider/awssd/aws_sd.go (hunk 15 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go:586]: still needed. I added a missing condition in the if

File: provider/awssd/aws_sd.go (hunk 16 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go]: it's the SRV endpoint that corresponds to the A endpoint. I replaced the keyword "preferred" with "corresponding"

File: provider/awssd/aws_sd.go (hunk 17 above)
Comment (SimonePomata) [reply provider/awssd/aws_sd.go:316]: `current.Targets` and `old.Targets` are always different, but they may have common items. For example, when a node is added to the k8s cluster, `old` contains all the existing nodes and `current` contains all the existing nodes plus the new one. This is an issue, because external-dns will try to de-register and then re-register the existing nodes (causing unneeded calls that overlap and result in errors). `DedupDeletesAndCreates` solves the issue by removing all items from `old` and by removing the same items from `current` (`current` then only contains the new node.).
